src/server/main_service.cc

```diff
@@ -1582,14 +1582,28 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
 
   interpreter->SetGlobalArray("KEYS", eval_args.keys);
   interpreter->SetGlobalArray("ARGV", eval_args.args);
-  interpreter->SetRedisFunc([cntx, this](auto args) { CallFromScript(cntx, args); });
 
   Interpreter::RunResult result;
   optional<ShardId> sid = GetRemoteShardToRunAt(*tx);
-  if (sid.has_value()) {
+  if (tx->GetMultiMode() != Transaction::NON_ATOMIC && sid.has_value()) {
     // If script runs on a single shard, we run it remotely to save hops.
-    pp_.at(sid.value())->Await([&]() { result = interpreter->RunFunction(eval_args.sha, &error); });
+    interpreter->SetRedisFunc([cntx, this](Interpreter::CallArgs args) {
+      // Disable squashing, as we're using the squashing mechanism to run remotely.
+      args.async = false;
+      CallFromScript(cntx, args);
+    });
+
+    boost::intrusive_ptr<Transaction> stub_tx = new Transaction{cntx->transaction};
+    cntx->transaction = stub_tx.get();
+    tx->PrepareSquashedMultiHop(registry_.Find("EVAL"), [&](ShardId id) { return id == *sid; });
+    tx->ScheduleSingleHop([&](Transaction* tx, EngineShard*) {
```

Comment:
please capture `stub_tx` by value

Reply:
actually, in this case it's not really needed.

Reply:
right :)